File: common/src/aws/sqs.rs
use aws_config::SdkConfig;
use aws_sdk_sqs::Client;

pub struct SqsClient {
  client: Client,
}

impl SqsClient {
  #[must_use]
  pub fn new(config: &SdkConfig) -> Self {
    let sqs_config = aws_sdk_sqs::config::Builder::from(config).build();
    SqsClient {
      client: Client::from_conf(sqs_config),
    }
  }

  pub async fn send_message(&self, queue_url: &str, message_body: &str) -> Result<(), String> {
    self
      .client
      .send_message()
      .queue_url(queue_url)
      .message_body(message_body)
      .send()
      .await
      .map_err(|e| format!("AWS SQS: SendMessage failed on Queue '{queue_url}' with error: {e}"))?;
    Ok(())
  }

  pub async fn receive_message(&self, queue_url: &str) -> Result<Option<(String, String)>, String> {
    let resp = self
      .client
      .receive_message()
      .queue_url(queue_url)
      .send()
      .await
      .map_err(|e| format!("AWS SQS: ReceiveMessage failed on Queue '{queue_url}' with error: {e}"))?;
    if let Some(message) = resp.messages.unwrap_or_default().into_iter().next() {
      let receipt_handle = message.receipt_handle.unwrap_or_default();
      let body = message.body.unwrap_or_default();
      Ok(Some((receipt_handle, body)))
    } else {
      Ok(None)
    }
  }

  pub async fn delete_message(&self, queue_url: &str, receipt_handle: &str) -> Result<(), String> {
    self
      .client
      .delete_message()
      .queue_url(queue_url)
      .receipt_handle(receipt_handle)
      .send()
      .await
      .map_err(|e| format!("AWS SQS: DeleteMessage failed on Queue '{queue_url}' with error: {e}"))?;
    Ok(())
  }

  pub async fn purge_queue(&self, queue_url: &str) -> Result<(), String> {
    self
      .client
      .purge_queue()
      .queue_url(queue_url)
      .send()
      .await
      .map_err(|e| format!("AWS SQS: PurgeQueue failed on Queue '{queue_url}' with error: {e}"))?;
    Ok(())
  }
}

#[cfg(test)]
mod tests {
  use crate::{aws, params};

  #[tokio::test(flavor = "multi_thread")]
  async fn test_sqs() {
    // Initialize an SQS client
    let sqs = aws::sqs::SqsClient::new(&params::AWS_SDK_CONFIG);
    let queue_url = "https://sqs.us-east-1.amazonaws.com/123456789012/my-queue"; // TODO: Put actual queue URL here

    // Test sending a message
    let send_result = sqs.send_message(queue_url, "Test message").await;
    assert!(send_result.is_ok());

    // Test receiving and deleting a message
    let receive_result = sqs.receive_message(queue_url).await;
    assert!(receive_result.is_ok());
    if let Some((receipt_handle, message)) = receive_result.unwrap() {
      println!("Received message: {}", message); // TODO: Compare to "Test message"
      let delete_result = sqs.delete_message(queue_url, &receipt_handle).await;
      assert!(delete_result.is_ok());
    }

    // Test purging the queue
    let purge_result = sqs.purge_queue(queue_url).await;
    assert!(purge_result.is_ok());
  }
}

File: common/aws/sqs.cpp
#include <aws/sqs/model/PurgeQueueRequest.h>
#include <aws/sqs/model/DeleteMessageRequest.h>
#include <aws/sqs/model/SendMessageRequest.h>
#include <aws/sqs/model/ReceiveMessageRequest.h>
#include "sqs.h"

AwsSQS::AwsSQS(const char* __restrict queueURL) : sqsClient(AWS::getSQSClient()), queueURL(queueURL) {}
AwsSQS::~AwsSQS(void) {}

bool AwsSQS::sendMessage(const char* __restrict message)
{
  // Send the specified message to the SQS queue
  Aws::SQS::Model::SendMessageRequest sendMessageRequest;
  sendMessageRequest.SetQueueUrl(queueURL);
  sendMessageRequest.SetMessageBody(message);
  const auto outcome = sqsClient->SendMessage(sendMessageRequest);
  if (!outcome.IsSuccess()) logger.log(Logger::ERROR, "AWS SQS: SendMessage failed on Queue '%s' with error: %s\n", queueURL.c_str(), outcome.GetError().GetMessage().c_str());
  return outcome.IsSuccess();
}

std::tuple<bool, std::string, std::string> AwsSQS::receiveMessage(int secondsToWait)
{
  // Receive a message from the SQS queue
  Aws::SQS::Model::ReceiveMessageRequest receiveMessageRequest;
  receiveMessageRequest.SetQueueUrl(queueURL);
  receiveMessageRequest.SetMaxNumberOfMessages(1);
  receiveMessageRequest.SetWaitTimeSeconds(secondsToWait);
  const auto outcome = sqsClient->ReceiveMessage(receiveMessageRequest);
  if (outcome.IsSuccess())
  {
    if (outcome.GetResult().GetMessages().empty())
    {
      logger.log(Logger::WARNING, "AWS SQS: No message received from Queue '%s' within %d seconds\n", queueURL.c_str(), secondsToWait);
      return std::make_tuple(false, "", "");
    }
    else
    {
      const auto& message = outcome.GetResult().GetMessages().front();
      return std::make_tuple(true, message.GetReceiptHandle(), message.GetBody());
    }
  }
  else
    logger.log(Logger::ERROR, "AWS SQS: ReceiveMessage failed on Queue '%s' with error: %s\n", queueURL.c_str(), outcome.GetError().GetMessage().c_str());
  return std::make_tuple(false, "", "");
}

bool AwsSQS::deleteMessage(const char* __restrict receiptHandle)
{
  // Delete the specified message from the SQS queue
  Aws::SQS::Model::DeleteMessageRequest deleteMessageRequest;
  deleteMessageRequest.SetQueueUrl(queueURL);
  deleteMessageRequest.SetReceiptHandle(receiptHandle);
  const auto outcome = sqsClient->DeleteMessage(deleteMessageRequest);
  if (!outcome.IsSuccess()) logger.log(Logger::ERROR, "AWS SQS: DeleteMessage failed on Queue '%s' with error: %s\n", queueURL.c_str(), outcome.GetError().GetMessage().c_str());
  return outcome.IsSuccess();
}

bool AwsSQS::purgeQueue(void)
{
  // Purge all messages from the specified SQS queue
  Aws::SQS::Model::PurgeQueueRequest purgeQueueRequest;
  purgeQueueRequest.SetQueueUrl(queueURL);
  const auto outcome = sqsClient->PurgeQueue(purgeQueueRequest);
  if (!outcome.IsSuccess()) logger.log(Logger::ERROR, "AWS SQS: PurgeQueue failed on Queue '%s' with error: %s\n", queueURL.c_str(), outcome.GetError().GetMessage().c_str());
  return outcome.IsSuccess();
}
